GitUI/Avatars/CustomAvatarProvider.UriTemplateData.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Security.Cryptography;
+using System.Text;
+using GitCommands;
+using JetBrains.Annotations;
+
+namespace GitUI.Avatars
+{
+    public sealed partial class CustomAvatarProvider
+    {
+        /// <summary>
+        /// A type that holds and prepares for variables in custom avatar templates.
+        /// It's processed by an <see cref="UriTemplateResolver"/>.
+        /// </summary>
+        private class UriTemplateData
+        {
+            public UriTemplateData([NotNull] string email, string name, int imageSize)
+            {
+                Email = email ?? throw new ArgumentNullException(nameof(email));
+                Name = name;
+                ImageSize = imageSize;
```

Comment:
Do we need to restrict size? E.g. to positive numbers, capped at a certain number.

Reply:
IMO that would lead to hidden/hard to detect bugs. We could throw an exceptions if it's something large or negative but just clamping it would only [hide or delay](https://en.wikipedia.org/wiki/Error_hiding) the underlying issue (namely that gitextensions somehow "thought" it was a good idea to request an avatar with a negative pixel size)

Reply:
There are number of types of errors though:
- those that a user can fix (e.g. user configuration related), and 
- those that the user is unable to fix. 

Also there's another dimension, e.g.:
- those errors that significant to the main operation (e.g. git is missing, git provides unexpected output, etc.), and
- those errors that may affect the UX but immaterial to the overall functionality.

E.g. if for some reason an avatar is requested with negative dimensions - we wouldn't want to crash the app because of it. It is certainly unfortunate to lose avatars, but this loss won't render the app useless.
Likewise, making a call to another system (i.e. a Gravatar, etc.) with data we know is invalid (i.e. dimensions are less than 1, or greater than, say, 300px) would be just as poor engineering practice as hiding exceptions.
The big question - if we end up in the situation with invalid dimensions, would the user be able to fix it themselves? Or should we mitigate the error by simply clamping the value to an expected range.

Reply:
There is now a safetynet that clamps sizes to a reasonable range, prevents negative sizes, and if something still fail provides an "emergency fallback" - a single pixel red dot. I thought about directly clamping sizes in various providers but this would just lead to a variety of edge cases in different places. Let me know what you think.